common/dbconfig.cpp

```diff
@@ -0,0 +1,103 @@
+// Copyright 2021-present Pinterest. All Rights Reserved.
+//
+// @author Premkumar (prem@pinterest.com)
+//
+
+#include "common/dbconfig.h"
+#include "common/segment_utils.h"
+#include <utility>
+
+#include <gflags/gflags.h>
+#include <glog/logging.h>
+
+DECLARE_int32(replicator_replication_mode);
+
+namespace common {
+
+DBConfigManager* DBConfigManager::get() {
+  static DBConfigManager instance;
+  return &instance;
+}
+
+DBConfigManager::DBConfigManager() : dbConfig(new DBConfig()) {
+}
+
+bool DBConfigManager::loadJsonStr(const std::string& content) {
+  Json::Value root;
+  Json::Reader reader;
+  if (!reader.parse(content, root) || !root.isObject()) {
+    LOG(ERROR) << "Could not load config";
+    return false;
+  }
+
+  return loadJsonObject(root);
+}
+
+bool DBConfigManager::loadJsonObject(const Json::Value& root) {
+  try {
+    auto newConfig = parseConfig(root);
+    if (newConfig == nullptr) {
+      return false;
+    }
+
+    LOG(INFO) << "Successfully updated config";
+    std::atomic_exchange_explicit(&dbConfig, newConfig, std::memory_order_release);
+    return true;
+  } catch (std::exception& ex) {
+    LOG(ERROR) << "Exception from parseConfig: " << ex.what();
+    return false;
+  }
+}
+
+std::shared_ptr<const DBConfig> DBConfigManager::parseConfig(const Json::Value& root) {
+  static const std::string ACK_MODE = "ack_mode";
+  static const std::string DATASET = "dataset";
+
+  DBConfig newDBConfig;
+
+  for (const auto& rootMember : root.getMemberNames()) {
+    // parse dataset specific config
+    if (rootMember == DATASET) {
+      auto dataSetConfigs = root[DATASET];
+
+      for (const auto& dataset : dataSetConfigs.getMemberNames()) {
+        Config config;
+        const auto& jsonDataSetConfig = dataSetConfigs[dataset];
+        bool isSet = false;
+        for (const auto& configName : jsonDataSetConfig.getMemberNames()) {
+          if (configName == ACK_MODE) {
+            config.replication_mode = jsonDataSetConfig[ACK_MODE].asUInt();
+            isSet = true;
+          }
+        }
+        if (isSet) {
+          newDBConfig.dataSet[dataset] = config;
+        }
+      }
+    }
+  }
+
+  return std::make_shared<const DBConfig>(std::move(newDBConfig));
+}
+
+/* conf getter functions */
+const Config* DBConfigManager::getConfig(const std::string& dbName) const {
+  auto segmentName = DbNameToSegment(dbName);
+  const auto dsIter = dbConfig->dataSet.find(segmentName);
+  if (dsIter == dbConfig->dataSet.end()) {
+    return nullptr;
+  } else {
+    return &(dsIter->second);
+  }
+}
+
+uint DBConfigManager::getReplicationMode(const std::string& dbName, uint defValue) const {
+  const auto* config = getConfig(dbName);
+  if (config != nullptr) {
+    return config->replication_mode;
```

Comment:
I think this could cause data race. 
When we write/swap the dbconfig, we are using `std::atomic_exchange_explicit(&dbConfig, newConfig, std::memory_order_release);` which is release.
Here, when we read it, we did not enforce any memory order, e.g. `acquire`. 
- without memory order, btw L86 and L96, another write for `dbConfig` could happen, the read/write will have no synchoronization, then, causing data race.

Reply:
Will fix this .